desc/coils.py

```diff
@@ -829,43 +853,74 @@ def flip(self, *args, **kwargs):
         basis : {"rpz", "xyz"}
-            basis for input coordinates and returned magnetic field
-        grid : Grid, int or None or array-like, optional
-            Grid used to discretize coil, the same for all coils. If an integer, uses
-            that many equally spaced points.
+            Basis for input coordinates and returned magnetic field.
+        source_grid : Grid, int or None or array-like, optional
+            Grid used to discretize coils. If an integer, uses that many equally spaced
+            points. Should NOT include endpoint at 2pi.
 
         Returns
         -------
         field : ndarray, shape(n,3)
-            magnetic field at specified points, in either rpz or xyz coordinates
+            Magnetic field at specified nodes, in [R,phi,Z] or [X,Y,Z] coordinates.
+
         """
+        assert basis.lower() in ["rpz", "xyz"]
+        if hasattr(grid, "nodes"):
+            grid = grid.nodes
+        grid = jnp.atleast_2d(grid)
         if params is None:
             params = [get_params(["x_s", "x", "s", "ds"], coil) for coil in self]
             for par, coil in zip(params, self):
                 par["current"] = coil.current
-        if hasattr(coords, "nodes"):
-            coords = coords.nodes
-        coords = jnp.atleast_2d(coords)
 
-        def body(B, x):
-            B += self[0].compute_magnetic_field(
-                coords, params=x, basis=basis, grid=grid
+        # stellarator symmetry is easiest in [X,Y,Z] coordinates
+        if basis.lower() == "rpz":
+            grid_xyz = rpz2xyz(grid)
+        else:
+            grid_xyz = grid
+
+        # if stellarator symmetric, add reflected nodes from the other half field period
+        if self.sym:
+            normal = jnp.array(
+                [-jnp.sin(jnp.pi / self.NFP), jnp.cos(jnp.pi / self.NFP), 0]
+            )
+            grid_sym = (
+                grid_xyz @ reflection_matrix(normal).T @ reflection_matrix([0, 0, 1]).T
             )
-            return B, None
+            grid_xyz = jnp.vstack((grid_xyz, grid_sym))
 
-        B = scan(body, jnp.zeros(coords.shape), tree_stack(params))[0]
+        # field period rotation is easiest in [R,phi,Z] coordinates
+        grid_rpz = xyz2rpz(grid_xyz)
 
+        # sum the magnetic fields from each field period
+        B = jnp.zeros_like(grid_rpz)
+        for k in range(self.NFP):
+            grid_nfp = grid_rpz + jnp.array([0, 2 * jnp.pi * k / self.NFP, 0])
+
+            def body(B, x):
+                B += self[0].compute_magnetic_field(
+                    grid_nfp, params=x, basis="rpz", source_grid=source_grid
+                )
+                return B, None
+
+            B += scan(body, jnp.zeros(grid_nfp.shape), tree_stack(params))[0]
+
+        # sum the magnetic fields from both halves of the symmetric field period
+        if self.sym:
+            B = B[: grid.shape[0], :] + B[grid.shape[0] :, :] * jnp.array([-1, 1, 1])
```

Comment:
TBH I don't understand why I have to subtract the B_R component instead of summing it like the B_phi and B_Z components. Maybe it would change to a sum with a different reflection for the symmetric nodes? But this seems to work.

Reply:
I think it makes sense. When you mirror the coil across the symmetry plane, you reverse all the field components. When you then flip it upside down you reverse Z and phi back to the correct way, but R is still backwards